src/L1/rollup/ScrollChain.sol

```diff
@@ -583,10 +623,83 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
         emit FinalizeBatch(_batchIndex, _batchHash, _postStateRoot, _withdrawRoot);
     }
 
+    /// @inheritdoc IScrollChain
+    function finalizeBundleWithTeeProof(
+        bytes calldata _batchHeader,
+        bytes32 _postStateRoot,
+        bytes32 _withdrawRoot,
+        bytes calldata _teeProof
+    ) external override OnlyProver whenFinalizeNotPaused {
+        // compute pending batch hash and verify
+        (uint256 batchPtr, bytes32 _batchHash, uint256 _batchIndex, ) = _loadBatchHeader(_batchHeader);
+
+        // check if this batch is finalized by zk proof.
+        bytes32 cachedStateRoot = finalizedStateRoots[_batchIndex];
+        if (cachedStateRoot == bytes32(0)) {
+            revert ErrorBatchNotFinalizedWithZKProof();
+        }
+
+        uint256 _finalizedBatchIndex = lastTeeFinalizedBatchIndex;
+        if (_batchIndex <= _finalizedBatchIndex) {
+            revert ErrorBatchIsAlreadyVerified();
+        }
+
+        bytes memory _publicInput = abi.encodePacked(
+            layer2ChainId,
+            uint32(_batchIndex - _finalizedBatchIndex), // numBatches
+            finalizedStateRoots[_finalizedBatchIndex], // _prevStateRoot
+            committedBatches[_finalizedBatchIndex], // _prevBatchHash
+            _postStateRoot,
+            _batchHash,
+            _withdrawRoot
+        );
+
+        uint256 batchVersion = BatchHeaderV0Codec.getVersion(batchPtr);
+        IRollupVerifier(teeVerifier).verifyBundleProof(batchVersion, _batchIndex, _teeProof, _publicInput);
+
+        // check withdraw root and state root match with zk proof
+        if (withdrawRoots[_batchIndex] != _withdrawRoot || cachedStateRoot != _postStateRoot) {
+            unresolvedState.batchIndex = _batchIndex;
+            unresolvedState.stateRoot = _postStateRoot;
+            unresolvedState.withdrawRoot = _withdrawRoot;
+            emit StateMismatch(_batchIndex, _postStateRoot, _withdrawRoot);
+            return;
+        }
+
+        lastTeeFinalizedBatchIndex = _batchIndex;
+
+        emit FinalizeBatchWithTEEProof(_batchIndex);
+    }
+
     /************************
      * Restricted Functions *
      ************************/
 
+    /// @notice Resolve mismatched state.
+    ///
+    /// @dev This should only be called by Security Council.
+    ///
+    /// @param useSGXState Whether we want to use the state root from SGX prover.
+    function resolveStateMismatch(bool useSGXState) external onlyOwner {
+        UnresolvedState memory state = unresolvedState;
+        if (state.batchIndex == 0) {
+            revert ErrorNoUnresolvedState();
+        }
+        if (useSGXState) {
+            finalizedStateRoots[state.batchIndex] = state.stateRoot;
+            withdrawRoots[state.batchIndex] = state.withdrawRoot;
```

Comment:
It seems to me that in this case we must reset `lastZkpVerifiedBatchIndex` (currently called `lastFinalizedBatchIndex`), since subsequent verified batches are invalidated.

Reply:
agreed, code added